### DISCUSSION TITLE
Source Generators on save

### DISCUSSION THREAD

**Discussion by YegorStepanov** [dotnet/roslyn#59388] [Q&A] [answered]
Is it possible to run source generators on save only, not real-time? 

The IDE experience degrades a lot with the frequent use of source generators

**Comment by sharwell** [comment, marked as answer]
> The IDE experience degrades a lot with the frequent use of source generators

We get this feedback often, but our observation over time is this is not true for cases where the source generator was optimally written using Roslyn's new APIs. The request to only run the source generator on save thus eventually reduces to a request to require users to opt-in to a work around for an inefficiently written source generator. It would be better to fix the source generator itself, thus not requiring a new user option _and_ not having the performance problem.

If you can provide more information about the goals of the source generator and the reason why you believe it is causing performance problems, we will likely be able to provide tips for correcting it.

**Reply by YegorStepanov** [reply]
I don't write them, I just use them.

> The request to only run the source generator on save thus eventually reduces to a request to require users to opt-in to a work around for an inefficiently written source generator. 

* Not everyone has the time to write a Roslyn-correct super-efficient generator.
* I'm not sure that the users want to think about caching, fast exits. 
* It will reduce the cost of writing it if you can do it in an easier way. So more users will write it.
* Some generators can be very heavy.
* The experience almost will be the same.
* I don't know any generator that should work on every symbol.

The [LoggerMessage] is added in .NET 6. Does its generator run on every symbol? Even when I write "Hello world" program?
Is the cost constant? 
If [LoggerMessage] generator needs 1000 CPU cycles for a small program using the attribute, would that also require 1000 cycles for a huge codebase?

**Reply by sharwell** [reply]
Logger message generator should only run if you change the signature of a logger method. It shouldn't run if you are just calling the logger method. We spent a good bit of time optimizing this one, and I haven't seen it show up as the cause of performance problems. We would certainly work to optimize it if we saw it showing up in [performance traces](https://docs.microsoft.com/en-us/visualstudio/ide/how-to-increase-chances-of-performance-issue-being-fixed?view=vs-2022#slowness-and-high-cpu-issues).

**Reply by YegorStepanov** [reply]
Oh, .NET previews suffered from SG v1. 

Since you had many reasons for adding the checkbox but decided not to, I'll close the question and wait for the libraries to be updated to SG v2.

> We get this feedback often

Did not find any discussion about it.

**Reply by sharwell** [reply]
>  and wait for the libraries to be updated to SG v2.

This work _should_ already be complete for source generators that ship in the SDK. If there are others that are still causing problems, please feel free to point them out to us so we can work with those authors to update them.

**Reply by YegorStepanov** [reply]
Thanks for the concern.

I use them in a home project:

__WindowsCommunityToolkit__: they've already moved to V2 but haven't updated the library yet.
__StrongInject__: want to add it but worried about IDE performance.

**Comment by xoofx** [comment]
I quite concur with the ask here. The default should have been to run a source generator only on save, not on every single key strokes. Even when using incremental source generator, it doesn't help when you see your generator running just because you changed a whitespace in a comment on the file and you are not done with your changes.

I don't know what is the real use case for realtime changes on every single key stroke, but I can't think of any for all the use cases I have been using it so far.

**Reply by xoofx** [reply]
I'm saying default, because this could be configurable at initialization time without much burden by selecting an interactivity mode e.g:

```csharp
public enum SourceGeneratorInteractivity {
    OnSave, // Default
    OnBuild, // build time only
    Realtime, // on each key strokes, not the default!
}
```

**Reply by xoofx** [reply]
In case you are wondering about the use cases: I made a [gist](https://gist.github.com/xoofx/ddcc713d941417de48524c200e74db14) one year ago to detect when a source generator is being used at build time, and to run it only from this situation. I have been redirecting multiple teams in my company to use this gist in order to reduce the impact these generators had on IDE experience.

**Reply by CyrusNajmabadi** [reply]
> I don't know what is the real use case for realtime changes

If you don't have realtime changes, then you'll end up getting tons of realtime errors as your code will spit out errors given its incompatibility with the code that hasn't been generated yet.  Requiring saving means you just sit there with errors that do not go away, despite your code being entirely correct, and totally compilable if you invoked a command line compile.

**Reply by CyrusNajmabadi** [reply]
> I have been redirecting multiple teams in my company to use this gist in order to reduce the impact these generators had on IDE experience.

It would be far better to just use incremental generators and just have low, or no, impact generators.  This is waht we did with all the SDK generators for 7.0 (and backported to 6.0).

**Reply by xoofx** [reply]
> If you don't have realtime changes, then you'll end up getting tons of realtime errors as your code will spit out errors given its incompatibility with the code that hasn't been generated yet. Requiring saving means you just sit there with errors that do not go away, despite your code being entirely correct, and totally compilable if you invoked a command line compile.

You are generalizing/abstracting on something that doesn't match the reality of many use cases I have seen: Take the example of Regex state machine generated from an attribute (or actually, almost any code generated behind a partial method). Do we need the state machine code to be generated on every key stroke? No. Do we need diagnostics of the validity of the regex string in realtime, probably yes.  This is the perfect example where the generated code is unnecessary at IDE time.

**Reply by mnmr** [reply]
> If you don't have realtime changes, then you'll end up getting tons of realtime errors as your code will spit out errors given its incompatibility with the code that hasn't been generated yet. Requiring saving means you just sit there with errors that do not go away, despite your code being entirely correct, and totally compilable if you invoked a command line compile.

I have a source generator that emits JsonConverters and type metadata (as C# models, since I can only generate C# but need the information to generate corresponding TypeScript code). It matters not one whit if this code is stale, all I need to ensure is that when the user builds the solution I generate new code. Forcing everyone to take an unnecessary performance hit in already struggling IDEs is a mindboggling decision imo.

**Reply by CyrusNajmabadi** [reply]
> Take the example of Regex state machine generated from an attribute (or actually, almost any code generated behind a partial method). Do we need the state machine code to be generated on every key stroke? 

As I said, we already fixed these generators. The regex generator will not do this anymore because it's properly incremental. It will only regenerate if something that affects it actually changes, which is precisely the behavior we want. 

Thank you for the example as it helps prove the point. We don't need things like "generate on save" because the solution was to make them properly incremental with the incremental generator apis.

**Reply by CyrusNajmabadi** [reply]
> Forcing everyone to take an unnecessary performance hit in already struggling IDEs is a mindboggling decision imo.

If your generator is properly incremental then no one should have a performance hit. Your converters and metadata should only regenerate when necessary *already*. 

Note: this is better than "compile on save" as it will not be a burden for users who save a lot (like myself). Because you are incremental, you won't have to pointlessly regenerate when nothing of interest is changing.

**Reply by Tornhoof** [reply]
But let's face it, writing such SG is a hard task, you even extended the logic just recently for more optimized SG via SyntaxValueProvider.ForAttributeWithMetadataName and poly filled it in runtime. 

Any 3rd party, community written SG is always behind with those advances. The dev experience around SG is already bad, writing them properly is complicated with many pitfalls to look out for, but those problems mostly don't impact the consumer side. Writing such SG using the more advanced functionality only makes it that much harder.

(disclaimer: I mostly gave up on writing SG, as the dev experience is simply not worth my time)

**Reply by CyrusNajmabadi** [reply]
> The dev experience around SG is already bad, 

If there are issues with the dev experience, please file them so they can be looked into.  Thanks!

**Reply by mnmr** [reply]
> If your generator is properly incremental then no one should have a performance hit. Your converters and metadata should only regenerate when necessary _already_.

It is not necessary for them to be regenerated every time the source changes. That is, in fact, completely unnecessary, and therefore constitutes a performance hit I'd rather be without.

> Note: this is better than "compile on save" as it will not be a burden for users who save a lot (like myself). Because you are incremental, you won't have to pointlessly regenerate when nothing of interest is changing.

My main interest is in an OnBuild option, which is not coupled to the save frequency. That said, if you can do an incremental generator that only does work when there are changes, the same system could be used in a save-triggered environment, so I don't really see your argument as being very strong here.

**Reply by CyrusNajmabadi** [reply]
> It is not necessary for them to be regenerated every time the source changes.

That's how an incremental generator already works. It only regenerates if something of relevance has changed and can this affect the final generated source.

**Reply by mnmr** [reply]
> > It is not necessary for them to be regenerated every time the source changes.
> 
> That's how an incremental generator already works. It only regenerates if something of relevance has changed and can this affect the final generated source.

I'll just assume that I am really bad at communicating, because as I see it that is absolutely not how it works.

Perhaps it'll help to list the two scenarios as I understand them:

Incremental generators:
1) Gets triggered on source changes
2) The generator has code to quickly exit if the modified symbol wasn't relevant
3) The generator has additional code to bypass source generation, or only do some of its work, deferring other parts for later
4) Eventually the later part gets to run to emit the final code
=> Impact on IDE: Some on every keystroke, some more when a relevant symbol is modified, some more when final code is emitted

Generate on build:
1) Scan symbol set in current compilation for relevant symbols
2) Emit code for those symbols
=> Impact on IDE: Only when building

These two scenarios are not the same. The first introduces a bunch of complexity into the generator, which for a lot of cases simply isn't needed. And it's not like source generators are non-complex to begin with, given that you have to learn both how they work and how to work with the Roslyn symbols.

**Reply by CyrusNajmabadi** [reply]
> Gets triggered on source changes

That's not really what happens. Roslyn is a 'pull' architecture. Features (of which there are literally hundreds, as well as whatever comes through from extensions) work on their own cadence and ask for semantic information. To get that accurate semantic information (which our apis have always provided), and which features (internal and external) ask for, that requires generation of SG content. 

However, incremental sgs can be smart. By seeing that the changes from the prior computation to this computation would not change results, they can avoid any generation actually happening.

**Reply by CyrusNajmabadi** [reply]
> Only when building

This is not a concept. There is no 'build' that pushes things along. E.g. we don't push parsing and semantics along with external events like building. We don't push them along with typing either. Instead, as mentioned above, it's all "pulling" and features (like analysis) request what they want. This is an extensible ecosystem as well (like SGs are) so analyzers do things like ask for correct information so that can give correct diagnostics.

**Comment by xoofx** [comment]
I have proposed an extension of this issue at #64173
